### ISSUE TITLE
A module in a role's library/ dir that overrides a built-in module is ignored if the module was used in an earlier role

### ISSUE THREAD

**Issue by agaffney** [ansible/ansible#17747]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

Loading of custom modules from a role
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```

and

```
ansible 2.2.0 (devel 1c7890bf86) last updated 2016/09/25 17:20:26 (GMT -600)
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

N/A
##### SUMMARY

When you override a built-in module with a version in a role's `library/` dir, that custom module will be ignored when the module has been used in another role from earlier in the play. The same does not happen when the module is used in `pre_tasks` in the play.
##### STEPS TO REPRODUCE

For this example, I placed a neutered copy of `stat.py` under `roles/foo/library/`. This version of the module always returns `{"changed": false, "stat": {"foo": "bar"}}` to make it obvious when it's being used.

File `roles/foo/tasks/main.yml`:

```
- stat: path=/tmp/foo
```

File `roles/bar/tasks/main.yml`:

```
- stat: path=/tmp/bar
```

Non-working playbook:

```
- hosts: localhost
  gather_facts: no
  roles:
    - bar
    - foo
```

Non-working output:

```
TASK [bar : stat] **************************************************************
ok: [localhost] => {"changed": false, "stat": {"exists": false}}

TASK [foo : stat] **************************************************************
ok: [localhost] => {"changed": false, "stat": {"exists": false}}
```

Working playbook:

```
- hosts: localhost
  gather_facts: no
  roles:
    #- bar
    - foo
```

Working output:

```
TASK [foo : stat] **************************************************************
ok: [localhost] => {"changed": false, "stat": {"foo": "bar"}}
```
##### EXPECTED RESULTS

The role `foo` should use its custom version of `stat.py`, regardless of whether the built-in version was used by a previous role.
##### ACTUAL RESULTS

The custom version of `stat.py` is NOT used.

**Comment by alikins** [comment]
This looks very similar to https://github.com/ansible/ansible/issues/17078

**Comment by alikins** [comment]
https://github.com/ansible/ansible/compare/devel...alikins:cache_invalid?expand=1  has a small changed that seemed to potentially fix #17078  (basically, invalid the plugin cache if we add paths to it)

**Comment by alikins** [comment]
Added a repro case for this bug at https://github.com/alikins/ansible-bug-repro/tree/master/role_custom_module_override_17747

compare/contrast with similar bug https://github.com/alikins/ansible-bug-repro/tree/master/role_custom_module_override_17078

**Comment by jimi-c** [comment]
@agaffney I'm going to push a fix to restore the functionality to 1.9.x and before's behavior (which in this case would be to always use the plugin found in the role). Note this would still use the incorrect plugin for the `bar` role in your example, however that has been a long-standing issue with roles. In the past, the guidance has always been to not override core/extras modules names as you're doing.

**Comment by jimi-c** [comment]
The commit to restore 1.9.x behavior is 5a57313dd72e574d38272364df3ef01d6a3646ef

**Comment by agaffney** [comment]
I was overriding a built-in module in a role because the built-in module had an issue that was fixed in the custom module version in the role. In my case, the custom version in the role being used in both roles is just fine.

**Comment by jimi-c** [comment]
Ok, per the above I'll close this now as it's merged into 2.2 (and devel).

Thanks!

**Comment by pcahyna** [comment]
Please reopen the issue, the fix was reverted in #17785

**Comment by abadger** [comment]
Isn't this the same as #17078
<https://github.com/ansible/ansible/issues/17078> ?  Maybe make sure that
@alikins test case in that big also covers this case?
